### ISSUE TITLE
Aead_ctx with given enc

### ISSUE THREAD

**Issue by GabrieleGalli** [rozbb/rust-hpke#33]
Hello,
I am using this library for my project and in particular agility;
I have four entities: 1 communicates with 2, 3 communicates with 4.
1 and 2 are very similar to your example (they have a common enc and exchange encrypted messages with aead_cxt sender and receiver respectively).
3 inherits a KDF(enc) from 1, 4 inherits a KDF(enc) from 2.
I would like to know if, and how, it is possible to create an aead_cxt_sender for 3 and an aead_cxt_receiver for 4 from these new enc_keys. In other words, is it possible to create aead_ctx using a given key, so that 3 and 4 can communicate?
Thanks in advance

**Comment by rozbb** [comment]
Hi Gabriele. So what you want is to allow 3 and 4 to derive an encryption context from 1 and 4, respectively? If so, then one thing you can do is have 1 and 3 [export](https://docs.rs/hpke/latest/hpke/aead/struct.AeadCtxR.html#method.export) a session secret. You can use this key as input to whatever secure messaging protocol you want to use. This crate does not provide such a protocol, though.

If you would like to set up a symmetric secure channel using a preshared key, one option is to set up a [STROBE](https://github.com/rozbb/strobe-rs) session, and `key()` it immediately. Does that help at all?

**Comment by GabrieleGalli** [comment]
Thank you for your reply, I tried what you suggested but it is still not what I was looking for. Let me explain in more detail.
Take for example 2 entities (Primary P, Secondary S).
I would like P.setup_sender to create a shared_secret and an encapped_key (which it already does) but not create an enc_cxt, so that this setup_sender would return Ok((encapped_key, shared_secret)). If this were possible at this point, I would like to use the shared_secret as input for S.setup_sender where, unlike P, I ONLY have to create an enc_cxt with the shared_secret as input. 
In other words, create a shared_secret in P, but create an enc_ctx in S. 
NOTICE THAT I'm using the agility, because I need to decide algorithms at run time.
Thank you in advance for your availability.

**Comment by rozbb** [comment]
I think I understand. This isn't directly possible in HPKE, and I don't want to add more methods to the API for a possibly uncommon use case. I still think my solution above would work, though. Is there a problem with the below?

```rust
let (encapped_key, unused_enc_ctx) = P.setup_sender(/* ... */);
let mut shared_secret = [0u8; 32];
unused_enc_ctx.export(&mut shared_secret);

// ... Over at S

ctx = Strobe::new(b"mysession", SecParam::B128);
ctx.key(&shared_secret, false);
let mut msg = b"hello world";
ctx.send_enc(&mut msg);

// msg is now encrypted. S sends it to someone else.
// Recipient gets msg:
ctx = Strobe::new(b"mysession", SecParam::B128);
ctx.key(&shared_secret, false);
ctx.recv_enc(&mut msg);
println!("Got {}", String::from_utf8_lossy(msg));
```

**Comment by GabrieleGalli** [comment]
Thank you very much! What I 'wanted' was an aead_cxt, but the Strobe works similarly; as I'm doing a proof of concept this is fine too, thanks again.
